### PR TITLE
Write tests, create inverted index class, implement front end template, automate with gulp, set up code coverage, coveralls, Travis CI. Some the main problem, and provide links to review, staging and production apps

### PR DESCRIPTION
**Pull request by desaawa**
#### What does this PR do?
-  Merges jasmine test, 
-  implement invertedIndex class, 
-  write front end template code,
-  Automate browser reloading with gulp,
-  Set up code coverage and coveralls,
-  Configure Travis CI
- Implement the main front end logic
- Deploy to heroku and provide links to review, staging and production apps.
- Final PR with Peer Code Reviews
- Reflection on learning experience
- Release notes for the product
- Explain two design patterns used

#### How should this be manually tested?
Clone the repo and install locally. Or view files changed in this PR
-  [Github Branch for implementing front end and automation with Gulp](https://github.com/andela-aawa/inverted-index/tree/feature/5/writing-html-css-js)
-  [Github Branch for test, test coverage, and Travis CI](https://github.com/andela-aawa/inverted-index/tree/feature/5/writing-html-css-js)
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
N/A
#### Screenshots (if appropriate)
image

#### Questions:PR Naming
N/A

### CODE REVIEW

**Comment by desaawa** [comment]
Hi @q-ode , I have made the changes as you requested on a new PR.

File: 06-building-a-program-start-to-finish/automating-tasks/output/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#### Output
+- **Using**: Gulp
+- **Deliver**: a screencast showing gulp watching and refreshing your browser when changes are made to your HTML, CSS & JS files
+- **With**: *independently*
+
+#### Link to Gulp Demo
+[Click Here](https://youtu.be/t6OAt-HnT1g)
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/automating-tasks/output/README.md]: All I see here is browser reload for HTML. Don't see anyone for JS and CSS.

File: 06-building-a-program-start-to-finish/checking-repository-health/output/README.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#### Output
+- **Using**: `Code Climate` and `Coveralls`
+- **Deliver**: a screenshot of your code coverage and GPA badges in your repo's README file
+- **With**: *independently*
+
+#### [Link to the Repo](https://github.com/andela-aawa/inverted-index/tree/chore/4/rewrite-tests)
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/checking-repository-health/output/README.md:6]: Great Job!

File: 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#### Output
+- **Using**: TravisCI
+- **Deliver**: a link to a successful TravisCI build of your `development` branch (it should include test runs)
+- **With**: *independently*
+
+#### [Link to Successful Travis Build](https://travis-ci.org/andela-aawa/inverted-index)
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md:6]: Even though this build passes, your latest build has failed. Please be mindful of this moving forward.

File: 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/output/README.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#### Output
+- **Using**: HTML, CSS & JS
+- **Deliver**: a folder containing the template for the proposed frontend of your project
+- **With**: *independently*
+
+#### Output
+[HTML and CSS for the app](https://github.com/andela-aawa/inverted-index/pull/18)
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/output/README.md:7]: Great Job!

The search however seems out of scope with the rest of the functionalities (i.e. Create Index, Upload).

It's also far away from the displayed result. It'll be too difficult to assume those functionalities go together.

File: 06-building-a-program-start-to-finish/writing-tests/output/README.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#### Output
+- **Using**: Mocha/Jasmine with any assertion library of choice (Chai.JS, Should.JS, Expect.JS...)
+- **Deliver**: a Pull Request showing all tests required for your project
+- **With**: *independently*
+
+#### Link to Pull Request for Tests
+- [Click here to view the PR](https://github.com/andela-aawa/inverted-index/pull/17)
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/writing-tests/output/README.md:7]: I've left a couple of comments on your PR.

Kindly review and revert.

**Review by q-ode — COMMENTED** [review]
Great Job!

Kindly address the concerns and give me an update when this is done.

File: 06-building-a-program-start-to-finish/automating-tasks/output/README.md (hunk 1 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/automating-tasks/output/README.md]: @q-ode I have made a new video. And changed the link.

File: 06-building-a-program-start-to-finish/checking-repository-health/output/README.md (hunk 2 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/checking-repository-health/output/README.md:6]: Thanks

File: 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md (hunk 3 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md:6]: @q-ode  I understand it is because I have not merged to the development branch yet. I needed you to make a review first.

File: 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/output/README.md (hunk 4 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/output/README.md:7]: Thanks.
I have moved the search closer.

File: 06-building-a-program-start-to-finish/writing-tests/output/README.md (hunk 5 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/writing-tests/output/README.md:7]: @q-ode I have made all changes as requested. Please have a look at them. I need to merge my branches and move on.

File: 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md (hunk 3 above)
Comment (q-ode) [reply 06-building-a-program-start-to-finish/setting-up-continuous-integration-(build-server)/output/README.md:6]: Ok, fine by me.

File: 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/output/README.md (hunk 4 above)
Comment (q-ode) [reply 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/output/README.md:7]: Nice!

File: 06-building-a-program-start-to-finish/solving-the-problem/output/README.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#### Output
+- **Using**: GitHub
+- **Deliver**: a Pull Request showing your implementation of the solution (e.g. InvertedIndex class)
+- **With**: *independently*
+
+#### Link to required Pull Request
+- [Here](https://github.com/andela-aawa/inverted-index/pull/17)
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/README.md:7]: Nice attempt!

I've put comments on the PR.

Kindly review and revert.

**Comment by desaawa** [comment]
Hi @q-ode. I have made changes based on the last review.
I have also completed and deployed the app to Heroku.

File: 06-building-a-program-start-to-finish/writing-tests/output/README.md (hunk 5 above)
Comment (q-ode) [reply 06-building-a-program-start-to-finish/writing-tests/output/README.md:7]: I don't see the files in this output folder as requested. Kindly fix.

Comment (desaawa) [reply 06-building-a-program-start-to-finish/writing-tests/output/README.md:7]: Okay.

File: 06-building-a-program-start-to-finish/solving-the-problem/output/README.md (hunk 6 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/README.md:7]: Okay. Changes can be found [here](https://github.com/andela-aawa/inverted-index/pull/20)

**Comment by desaawa** [comment]
Hi @q-ode. I have added the files in the output folders.

File: 06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/output/README.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#### Output
+- **Using**: Medium
+- **Deliver**: a post reflecting on your learning experience while producing these outputs
+- **With**: *independently*
+
+#### Link to Medium Post
+- Click [HERE](https://medium.com/@awa.desmoline/andela-simulations-checkpoint-one-eb472dc44f2f#.ot139btfo)
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/output/README.md:7]: Great Reflections!

Was an interesting read.

File: 06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/output/README.md [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#### Output
+- **Using**: Heroku
+- **Deliver**: live URLs to a `review` app, your `staging` & `production` apps
+- **With**: *independently*
+
+#### Links to `review`, `staging` & `production` apps
+- [Review](https://invertedindex-aawa-devel-pr-23.herokuapp.com/)
+- [Staging](https://invertedindex-aawa-develop.herokuapp.com/#)
+- [Production](https://invertedindex-aawa.herokuapp.com/)
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/output/README.md:9]: Great Job!

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
+/**
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Solid!

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  /**
+   * Create index
+   * @function
+   * @param {string} filename
+   * @param {Array} data
+   * @return {Object} index object
+   */
+  createIndex(filename, data) {
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Nitpick: What is `data`? #DescriptiveVariableNamesPlease

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
+/**
+ * InvertedIndex class
+ * @class
+ */
+class InvertedIndex {
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Please ensure your filename matches your class name. It's in the style guide.

File: 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js [hunk 12]
```diff
@@ -0,0 +1,139 @@
+const books = require('./books.json');
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js:1]: Please cleanup the style violations in this file.

**Review by q-ode — COMMENTED** [review]
Kindly address the comments.

File: 06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/output/README.md (hunk 7 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/output/README.md:7]: Thanks

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js (hunk 11 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Done

File: 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js (hunk 10 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/inverted-index.js]: Changed to `filedata`

File: 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js (hunk 12 above)
Comment (desaawa) [reply 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js:1]: Done. No more violations.

File: 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+const books2 = require('./books2.json');
+const books3 = require('./books3.json');
+
+describe('InvertedIndex Class', () => {
+  beforeAll(() => {
+    invertedIndex = new InvertedIndex();
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js]: Where was `invertedIndex` declared?

File: 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    });
+  });
+
+  describe('GetIndex', () => {
+    it('should return "undefined" if index does not exist', () => {
+      expect(invertedIndex.getIndex(' ')).toEqual(undefined);
+      expect(invertedIndex.getIndex('books4')).toEqual(undefined);
+    });
+    it('returns the exact result of the index', () => {
+      invertedIndex.createIndex('books3', books3);
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js]: Don't like the idea of there variable names `books3`...

File: 06-building-a-program-start-to-finish/using-design-patterns/output/README.md [hunk 15]
```diff
@@ -0,0 +1,54 @@
+#### Output
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/using-design-patterns/output/README.md:1]: Nice work!

**Review by q-ode — COMMENTED** [review]
Kindly address the two comments let's close this out!

File: 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js [hunk 16]
```diff
@@ -0,0 +1,137 @@
+const books = require('./books.json');
+const books2 = require('./books2.json');
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/writing-tests/output/inverted-index-test.js]: PS: I don't see you testing the functions used to validate the files being uploaded. Let's fix please.

**Comment by desaawa** [comment]
Hi, @q-ode, I have made required changes. Please take a look.

File: 06-building-a-program-start-to-finish/solving-the-problem/output/invertedIndex.js [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    this.indexes = {};
+  }
+
+  /**
+   * Validate File
+   * @function
+   * @param {string} filedata
+   * @return {Boolean} true or false
+   */
+  static validate(filedata) {
```
Comment (q-ode) [inline 06-building-a-program-start-to-finish/solving-the-problem/output/invertedIndex.js]: I believe the pattern is that all documents in each file being uploaded should contain `title` and `text` properties.

Comment (desaawa) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/invertedIndex.js]: Hi @q-ode, I have fixed it to text for both.

Comment (q-ode) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/invertedIndex.js]: @andela-aawa You're still only checking if the first document is valid. What if the file has 10 documents?

PS: Improve your JSDoc for this function.

Comment (desaawa) [reply 06-building-a-program-start-to-finish/solving-the-problem/output/invertedIndex.js]: @q-ode. I have made the fix. Sorry for the inconveniences.

**Review by q-ode — APPROVED** [review]
Awesome Job Man!

**Comment by desaawa** [comment]
Thanks @q-ode.
